aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -410,4 +418,38 @@ Tensor dot_cuda(const Tensor& self, const Tensor& other) {
   });
 }
 
+Tensor vdot_cuda(const Tensor& self, const Tensor& other) {
+  if (!self.is_complex()) {
+    return dot_cuda(self, other);
```

Comment:
```
>>> np.vdot(np.array([2]), np.array([2+3j]))
(4+6j)
>>> np.vdot(np.array([2+3j]), np.array([2]))
(4-6j)
```
maybe it's worth adding a comment that we only call `dot_cuda` when `self` is not complex because we want the above mentioned behavior.

Reply:
nvm I guess we throw an error for input tensors of different dtype:
```
>>> torch.dot(torch.tensor([1j]), torch.tensor([2]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dot : expected both vectors to have same dtype, but found ComplexFloat and Long
```
 I think we should add type promotion for `dot` and `vdot` to better align with numpy.

cc. @mruberry

Reply:
Type promotion would be nice but I think it's OK if it's not in this first PR. Type promotion is a little tricky to implement today when not using TensorIterator.

Reply:
Yeah. I was thinking about it. But since the plan is to use existing `dot` for real types, wasn't sure as to how to go about having type promotion.

Also other place where it diverges from numpy is that numpy operator supports broadcasting while `dot` doesn't. We can add broadcasting logic before passing the inputs to `dot`. What do you feel about it?

Reply:
I would file follow-up issues for broadcasting + type promotion. We may even want to add architecture to support doing these things easily outside of TensorIterator.

Reply:
numpy `vdot` does not support broadcasting, and we don't try to follow numpy `dot` behavior:
```
In [5]: np.vdot([2,3], [3])                                                                                                                                                                                                         
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-6192cfa76e6f> in <module>
----> 1 np.vdot([2,3], [3])

<__array_function__ internals> in vdot(*args, **kwargs)

ValueError: cannot reshape array of size 1 into shape (2,)
```

Reply:
@mruberry yeah that sounds good. out of curiosity -- what's tricky about implementing type promotion when not using TensorIterator? 

@kshitij12345  
```
if (!self.is_complex()) {
    return dot_cuda(self, other);
}
``` will give us the desired behavior for (real, complex) dot product when the type promotion is enabled, so we should still add a note documenting [that](https://github.com/pytorch/pytorch/pull/43004#discussion_r470631911). This is so that when the type promotion is enabled in future, we have the behavior already documented.

Reply:
Type promotion challenges:

- sometimes have to be careful to preserve your inputs (luckily not the case here)
- compute the result type
- cast inputs to the result type
- validate safe casting to out
- copy to out (if necessary)
- write a custom test for your op's type promotion behavior

It's not the end of the world. In this case we would want to change the behavior of dot, too, so it seems separable.